SwiftCompilerSources/Sources/Optimizer/Utilities/OwnershipUtils.swift

```diff
@@ -0,0 +1,184 @@
+//
+//   # lifetime introducer
+//   %1 = apply               -+                  -+
+//   ...                       | OSSA lifetime     |
+//   # forwarding instruction  |                   |
+//   %2 = struct $S (%1)      -+ -+                | forward-extended lifetime
+//                                | OSSA lifetime  |
+//   # non-forwarding consumer    |                |
+//   destroy_value %2            -+               -+
+//
+// The lifetime of a single owned value ends when it is forwarded, but certain lifetime properties are relevant for the entire forward-extended lifetime. For example, if an owned lifetime has a pointer-escaping use, then all values in the forward-extended lifetime are also considered pointer-escaping. Certain propeties, like lexical lifetimes, only exist on the forward introducer and apply to all forwarded values.
+//
+// Note: Although move_value conceptually forwards an owned value, it also summarizes lifetime attributes; therefore, it is not formally a ForwardingInstruction.
+//
+// TODO: when phi lifetime flags are implemented, phis will introduce a lifetime in the same way as move_value.
+//
+// The lifetime introducer of a guaranteed value is the borrow introducer:
+//
+//   # lifetime introducer / borrow introducer
+//   %1 = begin_borrow        -+
+//   ...                       | OSSA lifetime == forwarded lifetime
+//   # forwarding instruction  |
+//   %2 = struct $S (%1)       | - forwarded uses are within the OSSA lifetime
+//                             |
+//   end_borrow %1            -+
+//
+// TODO: When a begin_borrow has no lifetime flags, it can be ignored as a lifetime introducer. In that case, an owned value may introduce guaranteed OSSA lifetimes.
+//
+// This walker is used to query basic lifetime attributes on values, such as "escaping" or "lexical". It must be precise for correctness and is performance critical.
+protocol ForwardingUseDefWalker {
+  mutating func introducer(_ value: Value) -> WalkResult
+
+  // Minimally, check a ValueSet. This walker may traverse chains of aggregation and destructuring by default. Implementations may traverse phis.
+  mutating func needWalk(for value: Value) -> Bool
+
+  mutating func walkUp(value: Value) -> WalkResult
+}
+
+extension ForwardingUseDefWalker {
+  mutating func walkUp(value: Value) -> WalkResult {
+    walkUpDefault(value: value)
+  }
+  mutating func walkUpDefault(value: Value) -> WalkResult {
+    if let inst = value.definingInstruction as? ForwardingInstruction
+    {
+      return walkUp(instruction: inst)
+    }
+    return introducer(value)
+  }
+  mutating func walkUp(instruction: ForwardingInstruction) -> WalkResult {
+    for operand in instruction.forwardedOperands {
+      if needWalk(for: operand.value) {
+        if walkUp(value: operand.value) == .abortWalk {
+          return .abortWalk
+        }
+      }
+    }
+    return .continueWalk
+  }
+}
+
```

Comment:
coding style: in C++ we are used to put private types/functions _before_ the place where they are used. In swift this is not required. I tried to follow a top-down order: top-level entry points at the top and private types/functions following their uses.